ndpyramid/regrid.py

```diff
@@ -261,29 +269,154 @@ def pyramid_regrid(
         grid = target_pyramid[str(level)].ds.load()
         # get the regridder object
         if weights_pyramid is None:
-            regridder = xe.Regridder(ds, grid, method, **regridder_kws)
+            #regridder = xe.Regridder(ds, grid, method, **regridder_kws)
+            raise NotImplementedError("This requires xESMF, which we're trying to avoid")
         else:
             # Reconstruct weights into format that xESMF understands
             # this is a hack that assumes the weights were generated by
             # the `generate_weights_pyramid` function
 
             ds_w = weights_pyramid[str(level)].ds
             weights = _reconstruct_xesmf_weights(ds_w)
-            regridder = xe.Regridder(
-                ds, grid, method, reuse_weights=True, weights=weights, **regridder_kws
-            )
+        
         # regrid
         if regridder_apply_kws is None:
             regridder_apply_kws = {}
-        regridder_apply_kws = {**{'keep_attrs': True}, **regridder_apply_kws}
-        plevels[str(level)] = regridder(ds, **regridder_apply_kws)
+
+        plevels[str(level)] = xr_regridder(ds, grid, weights, out_grid_shape=(grid.sizes['x'], grid.sizes['y']))
+        
+        level_attrs = {
+            'multiscales': multiscales_template(
+                datasets=[{'path': '.', 'level': level, 'crs': projection_model._crs}],
+                type='reduce',
+                method='pyramid_regrid',
+                version=get_version(),
+                kwargs=save_kwargs,
+            )
+        }
+        plevels[str(level)].attrs['multiscales'] = level_attrs['multiscales']
 
     root = xr.Dataset(attrs=attrs)
     plevels['/'] = root
     pyramid = dt.DataTree.from_dict(plevels)
 
     pyramid = add_metadata_and_zarr_encoding(
-        pyramid, levels=levels, other_chunks=other_chunks, pixels_per_tile=pixels_per_tile, projection=Projection(name=projection)
+        pyramid,
+        levels=levels,
+        other_chunks=other_chunks,
+        pixels_per_tile=pixels_per_tile,
+        projection=Projection(name=projection),
     )
 
     return pyramid
+
+
+def xr_regridder(
+    ds: xr.Dataset,
+    grid: xr.Dataset,
+    weights: xr.DataArray,
+    out_grid_shape: tuple[int, int],
+) -> xr.Dataset:
+    """
+    Xarray-aware regridding function that uses weights from xESMF but performs the regridding using sparse matrix multiplication.
+    
+    Parameters
+    ----------
+    ds
+    weights
+    out_grid_shape
+
+    Returns
+    -------
+    regridded_ds
+    """
+
+    latlon_dims = ['nlat', 'nlon']
+
+    shape_in = (ds.sizes['nlat'], ds.sizes['nlon'])
+    shape_out = out_grid_shape
```

Comment:
This `out_grid_shape` argument is probably superfluous - you could just read it from the `grid` object.